pkg/avx/collector.go

```diff
@@ -71,117 +80,135 @@ var descriptors = [numDescriptors]*prometheus.Desc{
-func getElfKernelVersion(path string) (uint8, uint8, uint8, error) {
-	elfFile, err := elf.Open(path)
+func enablePerfTracepoint(prog *bpf.Program, tracepoint string) (int, error) {
+
+	id, err := ioutil.ReadFile(filepath.Join(kernelTracepointPath, tracepoint, "id"))
 	if err != nil {
-		return 0, 0, 0, errors.Wrapf(err, "unable to open ELF file %s", path)
+		return -1, errors.Wrap(err, "unable read tracepoint ID")
 	}
-	defer elfFile.Close()
 
-	sec := elfFile.Section("version")
-	if sec == nil {
-		return 0, 0, 0, errors.New("unable to find 'version' section")
-	}
-	data, err := sec.Data()
+	tid, err := strconv.Atoi(strings.TrimSpace(string(id)))
 	if err != nil {
-		return 0, 0, 0, errors.Wrap(err, "unable to get version data")
+		return -1, errors.New("unable to convert tracepoint ID")
 	}
 
-	// Least Significant Byte first
-	return data[2], data[1], data[0], nil
-}
+	attr := unix.PerfEventAttr{
+		Type:        unix.PERF_TYPE_TRACEPOINT,
+		Config:      uint64(tid), // tracepoint id
+		Sample_type: unix.PERF_SAMPLE_RAW,
+		Sample:      1,
+		Wakeup:      1,
+	}
 
-func checkElfKernelVersion(path string) error {
-	elfMajor, elfMinor, _, err := getElfKernelVersion(path)
+	pfd, err := unix.PerfEventOpen(&attr, -1, 0, -1, unix.PERF_FLAG_FD_CLOEXEC)
 	if err != nil {
-		return err
+		return -1, errors.Wrap(err, "unable open perf events")
 	}
 
-	currentCode, err := bpf.CurrentKernelVersion()
-	if err != nil {
-		return errors.Wrap(err, "unable to get current kernel version")
+	if _, _, errno := unix.Syscall(unix.SYS_IOCTL, uintptr(pfd), unix.PERF_EVENT_IOC_ENABLE, 0); errno != 0 {
+		return -1, errors.New("unable to set up perf events")
 	}
 
-	if currentCode < kernelVersionCode(elfMajor, elfMinor, 0) {
-		return errors.New("host kernel is too old, consider rebuilding eBPF")
+	if _, _, errno := unix.Syscall(unix.SYS_IOCTL, uintptr(pfd), unix.PERF_EVENT_IOC_SET_BPF, uintptr(prog.FD())); errno != 0 {
+		return -1, errors.New("unable to attach bpf program to perf events")
 	}
 
-	return nil
+	return pfd, nil
 }
 
-type collector struct {
-	root                     string
-	elfFilepath              string
-	bpfModule                *bpf.Module
-	avxContextSwitchCounters *bpf.Map
-	allContextSwitchCounters *bpf.Map
-	lastUpdateNs             *bpf.Map
-	lastCPUCounters          *bpf.Map
-}
+func getKernelVersion() uint32 {
 
-// NewCollector creates new Prometheus collector for AVX metrics
-func NewCollector() (prometheus.Collector, error) {
+	var uts syscall.Utsname
 
-	elfFilepath := path.Join(bpfInstallpath, bpfBinaryName)
+	err := syscall.Uname(&uts)
+	if err != nil {
+		return 0
+	}
 
-	if err := checkElfKernelVersion(elfFilepath); err != nil {
-		return nil, err
+	str := make([]byte, len(uts.Release))
+	for _, val := range uts.Release[:] {
+		if val == 0 {
+			break
+		}
+		str = append(str, byte(val))
 	}
 
-	bpfModule := bpf.NewModule(elfFilepath)
+	ver := strings.SplitN(strings.Trim(string(str), "\x00"), ".", 3)
```

Comment:
With the above `str` should be guarenteed to not have any null characters so you could drop `Trim()` for readability?